File: code/matrix_launcher/tools/auth_probe.cpp
#include <cstring>
#include <ctime>
#include <iostream>
#include <string>
#include <vector>

#include <winsock2.h>
#include <ws2tcpip.h>
#include <windows.h>

// ws2_32.lib is linked via -lws2_32 in the Makefile

#include "../matrixstaging/runtime/src/libltcrypto/auth_crypto.h"

namespace {

struct Options {
    std::string username;
    std::string password;
    int timeoutMs = 5000;
};

static bool ParseArgs(int argc, char** argv, Options* options) {
    if (!options) return false;

    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];

        if (arg == "--help" || arg == "-h") {
            std::cout << "Usage: " << argv[0] << " --username <name> [options]\n\n";
            std::cout << "Core CLI options:\n";
            std::cout << "  --username <name>\n";
            std::cout << "  --password <text>\n";
            std::cout << "  --timeout-ms <ms>\n";
            return false;
        } else if (arg == "--username") {
            if (++i >= argc) return false;
            options->username = argv[i];
        } else if (arg == "--password") {
            if (++i >= argc) return false;
            options->password = argv[i];
        } else if (arg == "--timeout-ms") {
            if (++i >= argc) return false;
            char* end = NULL;
            const unsigned long parsed = std::strtoul(argv[i], &end, 0);
            if (end == argv[i]) return false;
            options->timeoutMs = static_cast<int>(parsed);
        } else {
            std::cerr << "error: unknown option '" << arg << "'\n";
            return false;
        }
    }

    if (options->username.empty()) {
        std::cerr << "error: --username is required\n";
        return false;
    }
    return true;
}

static bool ConnectTcp(const std::string& host, uint16_t port, int timeoutMs, SOCKET* outFd) {
    if (!outFd) return false;
    *outFd = INVALID_SOCKET;

    WSADATA wsaData;
    const int wsaResult = WSAStartup(MAKEWORD(2, 2), &wsaData);
    if (wsaResult != 0) {
        std::cerr << "WSAStartup failed: " << wsaResult << "\n";
        return false;
    }

    struct addrinfo hints;
    std::memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;

    char portText[16] = {0};
    _snprintf_s(portText, sizeof(portText), _TRUNCATE, "%u", static_cast<unsigned>(port));

    struct addrinfo* results = NULL;
    const int gai = getaddrinfo(host.c_str(), portText, &hints, &results);
    if (gai != 0) {
        std::cerr << "getaddrinfo failed: " << gai_strerror(gai) << "\n";
        return false;
    }

    for (struct addrinfo* current = results; current; current = current->ai_next) {
        const SOCKET fd = socket(current->ai_family, current->ai_socktype, current->ai_protocol);
        if (fd == INVALID_SOCKET) continue;

        int opt = 1;
        setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (const char*)&opt, sizeof(opt));

        struct timeval tv;
        tv.tv_sec = timeoutMs / 1000;
        tv.tv_usec = (timeoutMs % 1000) * 1000;
        char timeout_buf[8];
        memcpy(timeout_buf, &tv, sizeof(tv));
        setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, timeout_buf, sizeof(tv));
        setsockopt(fd, SOL_SOCKET, SO_SNDTIMEO, timeout_buf, sizeof(tv));

        if (connect(fd, current->ai_addr, current->ai_addrlen) == 0) {
            *outFd = fd;
            freeaddrinfo(results);
            return true;
        }

        closesocket(fd);
    }

    freeaddrinfo(results);
    return false;
}

static bool SendAll(SOCKET fd, const std::vector<uint8_t>& bytes) {
    size_t offset = 0u;
    while (offset < bytes.size()) {
        const char* data_ptr = reinterpret_cast<const char*>(bytes.data() + offset);
        const ssize_t written = send(fd, data_ptr, static_cast<int>(bytes.size() - offset), 0);
        if (written < 0) {
            const int err = WSAGetLastError();
            if (err == WSAEWOULDBLOCK || err == WSAETIMEDOUT) continue;
            std::cerr << "send failed: " << err << "\n";
            return false;
        }
        offset += static_cast<size_t>(written);
    }
    return true;
}

static bool ReceiveExact(SOCKET fd, size_t byteCount, std::vector<uint8_t>* outBytes, bool* outTimedOut) {
    if (!outBytes || !outTimedOut) return false;
    *outTimedOut = false;
    outBytes->clear();

    while (outBytes->size() < byteCount) {
        uint8_t buffer[512];
        const size_t want = (byteCount - outBytes->size() < sizeof(buffer)) ? (byteCount - outBytes->size()) : sizeof(buffer);
        char* buffer_ptr = reinterpret_cast<char*>(buffer);
        const ssize_t got = recv(fd, buffer_ptr, static_cast<int>(want), 0);
        if (got == 0) return false;
        if (got < 0) {
            const int err = WSAGetLastError();
            if (err == WSAEWOULDBLOCK || err == WSAETIMEDOUT) {
                *outTimedOut = true;
                return false;
            }
            std::cerr << "recv failed: " << err << "\n";
            return false;
        }
        outBytes->insert(outBytes->end(), buffer, buffer + got);
    }
    return true;
}

static bool ReceivePacket(SOCKET fd, mxo::auth::FramedPacket* outPacket, bool* outTimedOut) {
    if (!outPacket || !outTimedOut) return false;

    std::vector<uint8_t> first;
    if (!ReceiveExact(fd, 1u, &first, outTimedOut)) return false;

    outPacket->headerBytes = first;
    size_t payloadLen = first[0];
    if (first[0] & 0x80u) {
        std::vector<uint8_t> second;
        if (!ReceiveExact(fd, 1u, &second, outTimedOut)) return false;
        outPacket->headerBytes.push_back(second[0]);
        payloadLen = ((static_cast<size_t>(first[0] & 0x7fu)) << 8u) | static_cast<size_t>(second[0]);
    }

    if (!ReceiveExact(fd, payloadLen, &outPacket->payloadBytes, outTimedOut)) return false;

    outPacket->bytes = outPacket->headerBytes;
    outPacket->bytes.insert(outPacket->bytes.end(), outPacket->payloadBytes.begin(), outPacket->payloadBytes.end());
    return true;
}

}  // namespace

int main(int argc, char** argv) {
    Options options;
    if (!ParseArgs(argc, argv, &options)) return 2;

    std::cout << "auth_probe config:\n";
    std::cout << "  host='auth.lith.thematrixonline.net'\n";
    std::cout << "  port=11000\n";
    std::cout << "  username='" << options.username << "'\n";
    std::cout << "  passwordProvided=" << (options.password.empty() ? 0 : 1) << "\n";
    std::cout << "  timeoutMs=" << options.timeoutMs << "\n\n";

    SOCKET fd = INVALID_SOCKET;
    if (!ConnectTcp("auth.lith.thematrixonline.net", 11000, options.timeoutMs, &fd)) {
        std::cerr << "error: connect failed\n";
        return 1;
    }
    std::cout << "connected\n\n";

    mxo::auth::FramedPacket getPublicKeyRequest;
    if (!mxo::auth::BuildGetPublicKeyRequestPacket(
            76005,  // launcher version
            0,       // current public key ID
            mxo::auth::kFrameModeAuto,
            &getPublicKeyRequest)) {
        std::cerr << "error: failed to build AS_GetPublicKeyRequest\n";
        closesocket(fd);
        return 1;
    }

    std::cout << "send 0x06 AS_GetPublicKeyRequest\n";
    if (!SendAll(fd, getPublicKeyRequest.bytes)) {
        closesocket(fd);
        return 1;
    }

    mxo::auth::FramedPacket getPublicKeyReplyPacket;
    bool timedOut = false;
    if (!ReceivePacket(fd, &getPublicKeyReplyPacket, &timedOut)) {
        std::cerr << "error: failed waiting for 0x07 reply" << (timedOut ? " (timeout)" : "") << "\n";
        closesocket(fd);
        return timedOut ? 0 : 1;
    }

    std::cout << "recv after 0x06\n\n";

    mxo::auth::GetPublicKeyReply publicKeyReply;
    if (!mxo::auth::ParseGetPublicKeyReplyPacket(
            getPublicKeyReplyPacket.bytes.data(),
            getPublicKeyReplyPacket.bytes.size(),
            &publicKeyReply)) {
        std::cerr << "error: failed to parse AS_GetPublicKeyReply\n";
        closesocket(fd);
        return 1;
    }

    std::cout
        << "parsed 0x07 fields:\n"
        << "  status=" << publicKeyReply.status << "\n"
        << "  currentTime=" << publicKeyReply.currentTime << "\n"
        << "  publicKeyId=" << publicKeyReply.publicKeyId << "\n"
        << "  keySize=" << static_cast<unsigned>(publicKeyReply.keySize) << "\n"
        << "  hasEmbeddedPublicKey=" << (publicKeyReply.hasEmbeddedPublicKey ? 1 : 0) << "\n\n";

    mxo::auth::AuthBlobLayout blobLayout;
    blobLayout.embeddedTime = static_cast<uint32_t>(std::time(NULL));
    blobLayout.twofishKey.clear();

    mxo::auth::AuthRequestLayout requestLayout;
    requestLayout.publicKeyId = publicKeyReply.publicKeyId;
    requestLayout.loginType = 1;
    requestLayout.keyConfigMd5.clear();
    requestLayout.uiConfigMd5.clear();

    mxo::auth::AuthRequestBuildResult authRequest;
    if (!mxo::auth::BuildAuthRequestPacket(
            options.username,
            blobLayout,
            requestLayout,
            mxo::auth::kFrameModeAuto,
            &authRequest)) {
        std::cerr << "error: failed to build AS_AuthRequest\n";
        closesocket(fd);
        return 1;
    }

    std::cout
        << "send 0x08 AS_AuthRequest\n"
        << "  authPublicKeyId=" << requestLayout.publicKeyId << "\n"
        << "  headerLen=" << authRequest.packet.headerBytes.size() << "\n"
        << "  payloadLen=" << authRequest.packet.payloadBytes.size() << "\n"
        << "  blobLen=" << authRequest.blobCiphertextBytes.size() << "\n\n";

    if (!SendAll(fd, authRequest.packet.bytes)) {
        closesocket(fd);
        return 1;
    }

    mxo::auth::FramedPacket postAuthReply;
    timedOut = false;
    if (!ReceivePacket(fd, &postAuthReply, &timedOut)) {
        if (timedOut) {
            std::cout << "recv after 0x08: timeout\n";
            closesocket(fd);
            return 0;
        }
        std::cout << "recv after 0x08: connection closed\n";
        closesocket(fd);
        return 1;
    }

    std::cout << "recv after 0x08\n\n";

    const uint8_t postAuthRawCode = postAuthReply.payloadBytes.empty() ? 0u : postAuthReply.payloadBytes[0];
    if (postAuthRawCode == 0x09u) {
        mxo::auth::AuthChallenge authChallenge;
        if (!mxo::auth::ParseAuthChallengePacket(
                postAuthReply.bytes.data(),
                postAuthReply.bytes.size(),
                &authChallenge)) {
            std::cerr << "error: failed to parse AS_AuthChallenge\n";
            closesocket(fd);
            return 1;
        }

        std::cout << "recv 0x09 AS_AuthChallenge\n";
        if (!options.password.empty()) {
            mxo::auth::AuthChallengeResponseLayout challengeResponseLayout;
            mxo::auth::AuthChallengeResponseBuildResult challengeResponse;
            if (!mxo::auth::BuildAuthChallengeResponsePacket(
                    authChallenge.encryptedChallengeBytes,
                    authRequest.twofishKeyBytes,
                    options.password,
                    "",  // soePassword (legacy, ignored)
                    challengeResponseLayout,
                    mxo::auth::kFrameModeAuto,
                    &challengeResponse)) {
                std::cerr << "error: failed to build AS_AuthChallengeResponse\n";
                closesocket(fd);
                return 1;
            }

            std::cout << "send 0x0A AS_AuthChallengeResponse\n";
            if (!SendAll(fd, challengeResponse.packet.bytes)) {
                closesocket(fd);
                return 1;
            }

            mxo::auth::FramedPacket authReplyPacket;
            timedOut = false;
            if (!ReceivePacket(fd, &authReplyPacket, &timedOut)) {
                std::cout << "recv after 0x0A: timeout\n";
                closesocket(fd);
                return timedOut ? 0 : 1;
            }

            std::cout << "recv after 0x0A\n\n";

            if (!authReplyPacket.payloadBytes.empty() && authReplyPacket.payloadBytes[0] == 0x0bu) {
                mxo::auth::AuthReply authReply;
                if (!mxo::auth::ParseAuthReplyPacket(
                        authReplyPacket.bytes.data(),
                        authReplyPacket.bytes.size(),
                        &authReply)) {
                    std::cerr << "error: failed to parse AS_AuthReply\n";
                    closesocket(fd);
                    return 1;
                }

                if (authReply.isErrorReply) {
                    std::cout
                        << "parsed 0x0B auth error:\n"
                        << "  errorCode=0x" << std::hex << authReply.errorCode
                        << " zeroDword=0x" << authReply.zeroDword
                        << " trailingWord=0x" << authReply.trailingWord
                        << std::dec << "\n";
                } else {
                    std::cout
                        << "parsed 0x0B auth success:\n"
                        << "  characterCount=" << authReply.characterCount << "\n"
                        << "  worldCount=" << authReply.worldCount << "\n"
                        << "  username='" << authReply.username.text << "'\n"
                        << "  offsetAuthData=0x" << std::hex << authReply.offsetAuthData
                        << " offsetEncryptedData=0x" << authReply.offsetEncryptedData
                        << " offsetCharData=0x" << authReply.offsetCharData
                        << " offsetServerData=0x" << authReply.offsetServerData
                        << std::dec << "\n";
                }
            }
        } else {
            std::cout << "challenge received, but no password provided\n";
        }
    } else if (postAuthRawCode == 0x0bu) {
        mxo::auth::AuthReply authReply;
        if (!mxo::auth::ParseAuthReplyPacket(
                postAuthReply.bytes.data(),
                postAuthReply.bytes.size(),
                &authReply)) {
            std::cerr << "error: failed to parse AS_AuthReply\n";
            closesocket(fd);
            return 1;
        }

        if (authReply.isErrorReply) {
            std::cout
                << "parsed 0x0B auth error:\n"
                << "  errorCode=0x" << std::hex << authReply.errorCode
                << " zeroDword=0x" << authReply.zeroDword
                << " trailingWord=0x" << authReply.trailingWord
                << std::dec << "\n";
        } else {
            std::cout
                << "parsed 0x0B auth success:\n"
                << "  characterCount=" << authReply.characterCount << "\n"
                << "  worldCount=" << authReply.worldCount << "\n"
                << "  username='" << authReply.username.text << "'\n";
        }
    }

    std::cout << "\n";
    closesocket(fd);
    return 0;
}

File: code/matrix_launcher/tools/auth_probe.py
#!/usr/bin/env python3
import argparse
import json
import pathlib
import socket
import struct
import sys
import time
from typing import Any

THIS_DIR = pathlib.Path(__file__).resolve().parent
sys.path.insert(0, str(THIS_DIR))
import build_authrequest_blob as auth_blob  # noqa: E402

AUTH_OPCODE_NAMES = {
    0x06: "AS_GetPublicKeyRequest",
    0x07: "AS_GetPublicKeyReply",
    0x08: "AS_AuthRequest",
    0x09: "AS_AuthChallenge",
    0x0A: "AS_AuthChallengeResponse",
    0x0B: "AS_AuthReply",
    0x35: "AS_GetWorldListRequest",
    0x36: "AS_GetWorldListReply",
}


def auth_opcode_name(opcode: int) -> str:
    return AUTH_OPCODE_NAMES.get(opcode, f"UNKNOWN_AUTH_0x{opcode:02X}")


def frame_payload(payload: bytes) -> bytes:
    payload_len = len(payload)
    if payload_len > 0x7FFF:
        raise ValueError("payload too large for current auth framing helper")
    if payload_len < 0x80:
        return bytes([payload_len]) + payload
    return bytes([0x80 | ((payload_len >> 8) & 0x7F), payload_len & 0xFF]) + payload


def recv_exact(sock: socket.socket, byte_count: int) -> bytes:
    chunks: list[bytes] = []
    remaining = byte_count
    while remaining > 0:
        chunk = sock.recv(remaining)
        if not chunk:
            raise ConnectionError(f"socket closed while waiting for {remaining} more byte(s)")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def recv_packet(sock: socket.socket) -> tuple[int, int, bytes]:
    first = recv_exact(sock, 1)[0]
    if first & 0x80:
        second = recv_exact(sock, 1)[0]
        payload_len = ((first & 0x7F) << 8) | second
        header_len = 2
    else:
        payload_len = first
        header_len = 1
    payload = recv_exact(sock, payload_len)
    return header_len, payload_len, payload


def parse_get_public_key_reply(payload: bytes) -> dict[str, Any]:
    if len(payload) < 14:
        raise ValueError("payload too short for AS_GetPublicKeyReply")
    if payload[0] != 0x07:
        raise ValueError(f"unexpected opcode 0x{payload[0]:02x} for GetPublicKeyReply")
    status, current_time, public_key_id = struct.unpack_from("<III", payload, 1)
    key_size = payload[13]
    tail = payload[14:]
    return {
        "opcode": payload[0],
        "message": auth_opcode_name(payload[0]),
        "status": status,
        "currentTime": current_time,
        "publicKeyId": public_key_id,
        "keySize": key_size,
        "tailHex": tail.hex(),
        "payloadHex": payload.hex(),
    }


def build_get_public_key_request(launcher_version: int, current_public_key_id: int) -> bytes:
    payload = bytes([0x06]) + struct.pack("<II", launcher_version, current_public_key_id)
    return frame_payload(payload)


def parse_md5_hex(text: str | None) -> bytes:
    if not text:
        return bytes(16)
    raw = bytes.fromhex(text)
    if len(raw) != 16:
        raise ValueError("MD5 override must decode to exactly 16 bytes")
    return raw


def build_auth_request_packet(
    public_key_id: int,
    username: str,
    login_type: int,
    key_config_md5: bytes,
    ui_config_md5: bytes,
    embedded_time: int | None,
    twofish_key_hex: str | None,
) -> tuple[bytes, dict[str, Any]]:
    twofish_key = auth_blob.parse_twofish_key(twofish_key_hex)
    blob_time = embedded_time if embedded_time is not None else int(time.time())
    plaintext = auth_blob.build_plaintext_blob(
        username=username,
        embedded_time=blob_time,
        twofish_key=twofish_key,
        rsa_method=4,
        some_short=0x1B,
    )
    ciphertext = auth_blob.encrypt_blob(plaintext)

    payload = bytearray()
    payload.append(0x08)
    payload += struct.pack("<I", public_key_id)
    payload.append(login_type & 0xFF)
    payload += b"\x00\x00"
    payload += key_config_md5
    payload += ui_config_md5
    payload += struct.pack("<H", len(ciphertext))
    payload += ciphertext

    framed = frame_payload(bytes(payload))
    metadata = {
        "publicKeyId": public_key_id,
        "loginType": login_type & 0xFF,
        "payloadLen": len(payload),
        "headerLen": len(framed) - len(payload),
        "byteCount": len(framed),
        "blobLen": len(ciphertext),
        "blobTime": blob_time,
        "twofishKeyHex": twofish_key.hex(),
        "plaintextBlobHex": plaintext.hex(),
        "ciphertextBlobHex": ciphertext.hex(),
    }
    return framed, metadata


def packet_summary(header_len: int, payload_len: int, payload: bytes) -> dict[str, Any]:
    opcode = payload[0] if payload else None
    return {
        "headerLen": header_len,
        "payloadLen": payload_len,
        "byteCount": header_len + payload_len,
        "opcode": opcode,
        "message": auth_opcode_name(opcode) if opcode is not None else "<empty>",
        "payloadHex": payload.hex(),
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Small auth-only launcher diagnostic probe")
    parser.add_argument("--host", default="auth.lith.thematrixonline.net")
    parser.add_argument("--port", type=int, default=11000)
    parser.add_argument("--username", required=True)
    parser.add_argument("--password", default="", help="Reserved for later challenge-response work")
    parser.add_argument("--launcher-version", type=lambda s: int(s, 0), default=76005)
    parser.add_argument("--current-public-key-id", type=lambda s: int(s, 0), default=0)
    parser.add_argument("--timeout", type=float, default=5.0)
    parser.add_argument("--send-authrequest", action="store_true")
    parser.add_argument("--login-type", type=lambda s: int(s, 0), default=1)
    parser.add_argument("--keyconfig-md5", default=None)
    parser.add_argument("--uiconfig-md5", default=None)
    parser.add_argument("--twofish-key-hex", default=None)
    parser.add_argument("--time", type=int, default=None)
    parser.add_argument("--json", action="store_true")
    return parser.parse_args()


def main() -> int:
    args = parse_args()

    results: dict[str, Any] = {
        "host": args.host,
        "port": args.port,
        "launcherVersion": args.launcher_version,
        "currentPublicKeyId": args.current_public_key_id,
        "username": args.username,
    }

    key_config_md5 = parse_md5_hex(args.keyconfig_md5)
    ui_config_md5 = parse_md5_hex(args.uiconfig_md5)

    with socket.create_connection((args.host, args.port), timeout=args.timeout) as sock:
        sock.settimeout(args.timeout)

        get_pub_packet = build_get_public_key_request(args.launcher_version, args.current_public_key_id)
        sock.sendall(get_pub_packet)
        results["sentGetPublicKeyRequest"] = {
            "rawCode": 0x06,
            "message": auth_opcode_name(0x06),
            "byteCount": len(get_pub_packet),
            "payloadLen": len(get_pub_packet) - 1 if get_pub_packet[0] < 0x80 else len(get_pub_packet) - 2,
            "packetHex": get_pub_packet.hex(),
        }

        header_len, payload_len, payload = recv_packet(sock)
        results["receivedAfterGetPublicKey"] = packet_summary(header_len, payload_len, payload)
        if payload and payload[0] == 0x07:
            results["parsedGetPublicKeyReply"] = parse_get_public_key_reply(payload)

        if args.send_authrequest:
            parsed_reply = results.get("parsedGetPublicKeyReply")
            if not parsed_reply:
                raise RuntimeError("cannot build AS_AuthRequest without a parsed AS_GetPublicKeyReply")

            auth_packet, metadata = build_auth_request_packet(
                public_key_id=int(parsed_reply["publicKeyId"]),
                username=args.username,
                login_type=args.login_type,
                key_config_md5=key_config_md5,
                ui_config_md5=ui_config_md5,
                embedded_time=args.time,
                twofish_key_hex=args.twofish_key_hex,
            )
            sock.sendall(auth_packet)
            results["sentAuthRequest"] = {
                "rawCode": 0x08,
                "message": auth_opcode_name(0x08),
                **metadata,
            }

            try:
                header_len2, payload_len2, payload2 = recv_packet(sock)
                results["receivedAfterAuthRequest"] = packet_summary(header_len2, payload_len2, payload2)
            except socket.timeout:
                results["receivedAfterAuthRequest"] = {
                    "timeout": True,
                    "message": "No reply within timeout after AS_AuthRequest",
                }

    if args.json:
        json.dump(results, sys.stdout, indent=2)
        sys.stdout.write("\n")
    else:
        print(json.dumps(results, indent=2))

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
